crates/turborepo-lib/src/process/child.rs

```diff
@@ -0,0 +1,473 @@
+pub struct Child {
+    pid: Option<u32>,
+    gid: Option<u32>,
+    state: Arc<RwLock<ChildState>>,
+    exit_channel: watch::Receiver<Option<i32>>,
+    stdin: Arc<Mutex<Option<tokio::process::ChildStdin>>>,
+    stdout: Arc<Mutex<Option<tokio::process::ChildStdout>>>,
+    stderr: Arc<Mutex<Option<tokio::process::ChildStderr>>>,
+}
+
+#[derive(Debug)]
+pub struct ChildCommandChannel(mpsc::Sender<ChildCommand>);
+
+impl ChildCommandChannel {
+    pub fn new() -> (Self, mpsc::Receiver<ChildCommand>) {
+        let (tx, rx) = mpsc::channel(1);
+        (ChildCommandChannel(tx), rx)
+    }
+
+    pub async fn kill(&self) -> Result<(), mpsc::error::SendError<ChildCommand>> {
+        self.0.send(ChildCommand::Kill).await
+    }
+
+    pub async fn stop(&self) -> Result<(), mpsc::error::SendError<ChildCommand>> {
+        self.0.send(ChildCommand::Stop).await
+    }
+}
+
+pub enum ChildCommand {
+    Stop,
+    Kill,
+}
+
+impl Child {
+    /// Start a child process, returning a handle that can be used to interact
+    /// with it. The command will be started immediately.
+    pub fn spawn(mut command: Command, shutdown_style: ShutdownStyle) -> Self {
+        let mut group = command.group().spawn().expect("failed to start child");
+
+        let stdin = group.inner().stdin.take();
+        let stdout = group.inner().stdout.take();
+        let stderr = group.inner().stderr.take();
+
+        let (command_tx, mut command_rx) = ChildCommandChannel::new();
+        let (exit_tx, exit_rx) = watch::channel(None);
+
+        let state = Arc::new(RwLock::new(ChildState::Running(command_tx)));
+        let task_state = state.clone();
+        let pid = group.inner().id();
+        let gid = group.id();
+
+        let mut child = group.into_inner();
+
+        let _task = tokio::spawn(async move {
+            info!("waiting for task");
+            tokio::select! {
+                command = command_rx.recv() => {
+                    let state = match command {
+                        Some(ChildCommand::Stop) | None => {
+                            // we received a command to stop the child process
+                            shutdown_style.process(&mut child, Some(exit_tx)).await.unwrap()
+                        }
+                        Some(ChildCommand::Kill) => {
+                            // we received a command to kill the child process
+                            debug!("killing child process");
+                            ShutdownStyle::Kill.process(&mut child, Some(exit_tx)).await.unwrap()
+                        }
+                    };
+
+                    {
+                        let mut task_state = task_state.write().await;
+                        *task_state = state;
+                    }
+                }
+                status = child.wait() => {
+                    // the child process exited
+                    let exit_code = status.ok().and_then(|s| s.code());
+                        {
```

Comment:
Nit: I think the indentation here is weird? 

Also, in what scenarios can `child.wait()` fail? It looks like we're throwing away the error and just sending `None` on the channel...

Reply:
I have folded that error into the return type